tsdb/engine/tsm1/ring.go

```diff
@@ -0,0 +1,289 @@
+package tsm1
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"sync/atomic"
+
+	"github.com/cespare/xxhash"
+)
+
+// ring is a structure that maps series keys to entries.
+//
+// ring is implemented as a crude hash ring, in so much that you can have
+// variable numbers of members in the ring, and the appropriate member for a
+// given series key can always consistently be found. Unlike a true hash ring
+// though, this ring is not resizeable—there must be at most 256 members in the
+// ring, and the number of members must always be a power of 2.
+//
+// ring works as follows: Each member of the ring contains a single store, which
+// contains a map of series keys to entries. A ring always has 256 partitions,
+// and a member takes up one or more of these partitions (depending on how many
+// members are specified to be in the ring)
+//
+// To determine the partition that a series key should be added to, the series
+// key is hashed and the first 8 bits are used as an index to the ring.
+//
+type ring struct {
+	// The unique set of partitions in the ring.
+	// len(partitions) <= len(continuum)
+	partitions []*partition
+
+	// A mapping of partition to location on the ring continuum. This is used
+	// to lookup a partition.
+	continuum []*partition
+
+	// Number of keys within the ring. This is used to provide a hint for
+	// allocating the return values in keys(). It will not be perfectly accurate
+	// since it doesn't consider adding duplicate keys, or trying to remove non-
+	// existent keys.
+	keysHint int64
+}
+
+// newring returns a new ring initialised with n partitions. n must always be a
+// power of 2, and for performance reasons should be larger than the number of
+// cores on the host. The supported set of values for n is:
+//
+//     {1, 2, 4, 8, 16, 32, 64, 128, 256}.
+//
+func newring(n int) (*ring, error) {
+	if n <= 0 || n > 256 {
```

Comment:
Also, we should probably test w/ higher partitions under high load to see if that has any benefit.